pages/workspace/workspace.html.js

```diff
@@ -288,11 +316,39 @@ function Workspace({ navTitle, showLogs, resLocals }) {
 
             socket.emit('startWorkspace', { workspace_id: workspaceId });
 
-            setInterval(() => {
-              socket.emit('heartbeat', { workspace_id: workspaceId }, (msg) => {
-                console.log('heartbeat, msg =', msg);
-              });
-            }, heartbeatInterval * 1000);
+            let heartbeatIntervalId = null;
+
+            function startHeartbeat() {
+              if (heartbeatIntervalId) return;
+
+              heartbeatIntervalId = setInterval(() => {
+                socket.emit('heartbeat', { workspace_id: workspaceId }, (msg) => {
+                  console.log('heartbeat, msg =', msg);
+                });
+              }, heartbeatIntervalSec * 1000);
```

Comment:
Say you have a hidden workspace page that you switch to for 30 seconds and then switch away. When we switch to the page we will start the heartbeat timer, but `setInterval` will wait 60 seconds (`heartbeatIntervalSec`) before it sends the first heartbeat so it won't send one before we stop the timer again. This means we will only send a heartbeat if we actually stay on the workspace page for at least 60 seconds continuously.

```suggestion
              function sendHeartbeat() {
                socket.emit('heartbeat', { workspace_id: workspaceId });
              }
              sendHeartbeat(); // send immediately
              heartbeatIntervalId = setInterval(sendHeartbeat, heartbeatIntervalSec * 1000);
```

The above solution is simple but it has the downside that we'll send lots of heartbeats if the user is constantly toggling back and forwards between a workspace page and another page.

Another possible solution would be something like:
```
let pageLastVisibleTime = Date.now();

setInterval(() => {
  if (Document.visibilityState == 'visible') {
    // Update our page visibility record if we are currently visible.
    pageLastVisibleTime = Date.now();
  }
  // Only send a heartbeat if this page was recently visible.
  if (Date.now() - pageLastVisibleTime < visibilityTimeoutSec * 1000) {
    socket.emit('heartbeat', { workspace_id: workspaceId });
  }
}, heartbeatIntervalSec * 1000);

document.addEventListener('visibilitychange', () => {
  // Every time we switch to or from this page, record that it was visible.
  // This is needed to capture the visibility when we switch to this page
  // and then quickly switch away again.
  pageLastVisibleTime = Date.now();
});
```

The above solution has the disadvantage that the client will still use resources to wake up every minute and check whether the page is visible, but I think this is pretty minimal usage on the client. The main advantages of this solution are that the code is simple with no start/stop of timers, and that it minimizes load on the server because clients will only send at most one heartbeat every 60 seconds.

Reply:
I like the second solution better - I'm all about minimizing server load! I'll make that change.

Reply:
Done in bac0689.

Reply:
Seems to be missing the `visibilitychange` handler from the code above?